[#14] Member 클래스의 Getter에 의한 고객 비밀번호 노출 문제 해결 1차 시도

choichanhyeok opened the pull request:
## a. 설명
> [#14] 이슈 해소를 위해 Member 클래스에 `@Getter`를 제거함. 이로인해 `member.getpassword()` 을 사용하던 `SupermarketUserDetailsService`에서 로그인시 인증/인가를 위해 필요한 `userdetails.UserDetails`을 내려주지 못하게됬는데 이 문제를 해결하기 위해 Member 객체에서 `userdetails.UserDetails`를 내려주는 방식을 적용해 getter 없이도 기존 로직이 동작토록 함.

## b. 작업 내용
> 1. Member 클래스에 `@Getter` 제거
```
비밀번호 및 고객ID 조회 이슈를 막기 위해 @Getter 자체를 제거
```

> 2. SupermarketUserDetailsService에서 인증 절차를 위해 필요했던 `userdetails.UserDetails`를 내려주는 역할을 member에게 양도
```
멤버에서 userdetails.UserDetails를 상속받는 User 객체를 직접 만들어 내려주는 toUserDetails() 작성
```

> 3. Member의 비밀번호 조회를 위한 2차 보안성 검토
```
멤버로 부터 전달받은 User 객체를 통해 멤버의 비밀번호를 조회가능한 취약점 발견 및 이를 문서화하기 위한 테스트 코드 작성
```
<img width="430" alt="유저도 결국 " src="https://github.com/f-lab-edu/super-market/assets/68278903/1134cbdb-ce03-4dd9-a310-229fb709b6e8">

## c. 작업 회고
> 쉽게 해결할 수 있는 문제인 줄 알았는데 생각보다 쉽지 않았다. 추가적으로 발생한 User의 getPassword에 의한 비밀번호 조회 문제는 `userdetails.User`를 상속받는 클래스를 만들고 `getPassword()` 부분만 재정의 하고 확인해볼까 싶었는데 스프링 시큐리티의 동작 방식에 대해 잘 이해하지 못한 상태에서 이런 접근 방식은 의미가 없을 거 같아 일단 멤버 클래스를 통해 직접 조회하는 걸 막는 부분까지만 진행했다.

## d. 기타 사항
> [#12]의 이슈에 대한 PR 전이어서 어쩔 수 없이 패키지 명 수정을 같이 해주었습니다. 읽으시는데 번거로움 만들어 죄송합니다!

choichanhyeok committed:
멤버 클래스의 Getter 제거 및 userDetails와의 연동을 유지하기 위해 UsertoUserDetails() 작성

- getter에 의한 보안 이슈 확인 [#14]를 해소하기 위해 @Getter 제거
- 인증/인가 과정에 필요한 userdetails.User 객체를 Getter 없이 생성키 위해 toUserDetails() 작성

choichanhyeok committed:
SupermarketUserDetailsService에서 `userdetails.User`을 생성하는 방식 변경

- SupermarketUserDetailsService에서 인증/인가를 위해 사용하는 Member 클래스의 getter가 [#14]이슈 해소를 위해 사라짐.
 - 기존 기능을 유지하기 위해 고객정보 기반 User 클래스 생성을 대체하는 toUserDetails()메서드 적용

choichanhyeok committed:
[#14] 이슈 해소를 위해 Member 클래스의 getter 제거 이후 비밀번호 조회 여부에 대한 2차 보안성 검토 진행

- member의 getter를 이용해 직접 조회는 불가
- 하지만 반환되는 userDetials를 이용한 getter를 통한 비밀번호 조회가 가능하다는 취약점 발견
- 이에 대한 해결 방안은 차후 검토 예정

f-lab-troy commented on supermarket-api-gateway/src/test/java/com/harmony/supermarketapigateway/auth/entity/MemberTest.java:
```diff
@@ -0,0 +1,29 @@
+
+class MemberTest {
+    @Nested
+    class 멤버_객체에_대한_보안성_체크 {
```
class, method는 영어로만 사용하시길 추천드립니다.

choichanhyeok replied:
위 피드백에 대한 내용 [수정 완료](https://github.com/f-lab-edu/super-market/pull/15/commits/dbbf302864cbda0aef19aaaf1f0cdcc26ccf357c) 했습니다. 수정에 대한 의사결정 기준은 아래와 같습니다!

<br>

> 1. 테스트에 대한 표현은 영어로 해야하는가
- 해당 프로젝트를 작업하는 작업자가 한국어에 더 익숙해 한국말로 표현키로 했다.
- 클래스, 메서드 명을 영어로 표기하는걸 권고하신다는 피드백도 같이 적용키 위해 `@DisplayName`을 적용키로 했다.

> 2. 테스트에 대한 메서드나 클래스에 의미 있는 이름이 부여되야 하는가
- 어차피 DisplayName을 통해 충분히 해당 메서드가 설명되는데 test1(), test2() 같은 메서드 명으로 편리하게 써주어도 되지 않을까 생각이 들었다.
- 다시 생각해보니 메서드 이름을 짓는데 드는 시간이 그렇게 많이 소요되는 일이 아니라 굳이 메서드 이름을 간소화해서 작성할 필요는 없을 거 같다는 생각이 들었다.

choichanhyeok committed:
한글로 작성되었던 테스트를 위한 클래스, 메서드 명을 영어로 변경

- `멤버_객체에_대한_보안성_체크` -> `멤버_객체에_대한_보안성_체크`
- `getter와_toString에_의한_비밀번호_노출_된다` -> `testUserDetailsExposePassword`
- 그 외 변수명 및 주석을 이해하기 쉽게 변경
